File: template/ufs/ufs.go
package ufs

type Ufs []int

func New(n int) Ufs {
	ufs := make([]int, n)
	for i := 0; i < n; i++ {
		ufs[i] = -1
	}
	return ufs
}

func (ufs Ufs) Find(x int) int {
	if px := ufs[x]; px < 0 {
		return x
	} else {
		ufs[x] = ufs.Find(px)
		return ufs[x]
	}
}

func (ufs Ufs) Size(x int) int {
	px := ufs.Find(x)
	return -ufs[px]
}

func (ufs Ufs) Union(x, y int) {
	px := ufs.Find(x)
	py := ufs.Find(y)
	if px != py {
		ufs[px] += ufs[py]
		ufs[py] = px
	}
}

File: src/main/java/template/Ufs.java
package template;

public class Ufs {

    private int[] ufs;

    public Ufs(int n) {
        ufs = new int[n];
        for (int i = 0; i < n; i++) {
            ufs[i] = -1;
        }
    }

    public void union(int i, int j) {
        int pi = find(i);
        int pj = find(j);
        if (pi != pj) {
            ufs[pi] += ufs[pj];
            ufs[pj] = pi;
        }
    }

    public int size(int i) {
        int pi = find(i);
        return -ufs[pi];
    }

    public int find(int i) {
        if (ufs[i] < 0) {
            return i;
        } else {
            ufs[i] = find(ufs[i]);
            return ufs[i];
        }
    }

}
